lib/engine/game/g_18_ts/game.rb

```diff
@@ -0,0 +1,1166 @@
+          raise OptionError, 'Cannot use K1 or K6 with D1' if (@kits[1] || @kits[6]) && optional_rules.include?(:d1)
+          if !units[1] && !units[3] && optional_rules.include?(:db1)
+            raise OptionError, 'Variant DB1 not useful in a Unit 2 only game'
+          end
+          raise OptionError, 'Variant DB2 is for Unit 1' if !units[1] && optional_rules.include?(:db2)
+          raise OptionError, 'Variant DB3 is for Unit 3' if !units[3] && optional_rules.include?(:db3)
+          raise OptionError, 'Unit 4 requires Unit 3' if optional_rules.include?(:unit_4) && !units[3]
+
+          p_range = case @units.keys.sort.map(&:to_s).join
+                    when '1'
+                      [3, 7]
+                    when '2'
+                      [3, 5]
+                    when '3'
+                      [4, 9]
+                    end
+          if (p_range.first > @players.size || p_range.last < @players.size) && @players.size.positive?
+            raise OptionError, 'Invalid option(s) for number of players'
+          end
+
+          optional_rules
+        end
+
+        def calculate_bank_cash
+          bank_cash = 0
+          if @optional_rules.include?(:big_bank)
+            bank_cash += BANK_UNIT1 if @units[1]
+            bank_cash += BANK_UNIT2 if @units[2]
+            bank_cash += BANK_UNIT3 if @units[3]
+          else
+            bank_cash = BANK_UNIT1 if @units[1]
+            bank_cash = BANK_UNIT2 if @units[2] && bank_cash.zero?
+            bank_cash = BANK_UNIT3 if @units[3] && bank_cash.zero?
+          end
+
+          # add in minor and kit changes (Mike Hutton clarification)
+          unless @optional_rules.include?(:strict_bank)
+            bank_cash += 2000 if @kits[2]
+            bank_cash += 2000 if @kits[3]
+            bank_cash += 1000 * num_minors
+          end
+
+          bank_cash
+        end
+
+        def bank_by_options
+          @bank_by_options ||= calculate_bank_cash
+        end
+
+        def cash_by_options
+          case @units.keys.sort.map(&:to_s).join
+          when '1'
+            { 2 => 1260, 3 => 840, 4 => 630, 5 => 504, 6 => 420, 7 => 360 }
+          when '2'
+            { 2 => 1260, 3 => 840, 4 => 630, 5 => 504 }
+          when '3'
+            { 4 => 630, 5 => 504, 6 => 420, 7 => 360, 8 => 315, 9 => 280 }
+
+          end
+        end
+
+        def num_minors
+          num_minors = 0
+          num_minors += 2 if @regionals[1]
+          num_minors += 1 if @regionals[2]
+          num_minors += 1 if @regionals[2]
+          num_minors += 2 if @kits[5]
+          num_minors += 1 if @kits[7]
+          num_minors
+        end
+
+        def adjust_certs(certs, chash)
+          chash.keys.each do |nplayers|
+            if certs[nplayers]
+              certs[nplayers] += chash[nplayers]
+            else
+              certs[nplayers] = chash[nplayers]
+            end
+          end
+        end
```

Comment:
This method isn't called anywhere in this pull request, so is it needed?

If it is called elsewhere, it would be better to use a hash with a default value. `certs = Hash.new(0)` gives you a hash where you can later write `certs[5] += 2` and it will either add two to the existing value, or create a new entry with a value of zero and then add two to it.